tcmur_device.c

```diff
@@ -156,7 +156,7 @@ void tcmu_notify_conn_lost(struct tcmu_device *dev)
 	struct tcmur_device *rdev = tcmu_get_daemon_dev_private(dev);
 
 	pthread_mutex_lock(&rdev->state_lock);
-	if (rdev->flags & TCMUR_DEV_FLAG_IN_RECOVERY)
+	if ((rdev->flags & TCMUR_DEV_FLAG_STOPPING) || (rdev->flags & TCMUR_DEV_FLAG_IN_RECOVERY))
 		goto unlock;
```

Comment:
There are 2 checks for STOPPING in __tcmu_reopen_dev which is called a little later by the thread started by this function. Why are you not hitting those checks?

Without this patch, do you get stuck in a handler->open/close call for extended periods (minutes) and it is looking like a hang?

Reply:
@mikechristie 
Test again without this patch, and this log reveals more details:
1. cmd timeout arrives
2. nl msg arrives
3. tcmu_add_dev_to_recovery_list failed to start recovery thread due to  enabled_port = 0
and returned with ret=0, which will cause TCMUR_DEV_FLAG_IN_RECOVERY wrongly set. 
('enabled_port = 0' is wired because I created my device with alua_access_type=1, need your advise here.) whatever TCMUR_DEV_FLAG_IN_RECOVERY wrongly set is a bug when enabled_port = 0.

# tcmu_add_dev_to_recovery_list()
{
...
        if (!enabled_port) {
                /* User disabled port from under us? */
                tcmu_dev_err(dev, "Could not find enabled port.\n");
                goto done;
        }  
...
done:
        tcmu_release_alua_grps(&alua_list);
        pthread_mutex_unlock(&tpg_recovery_lock);
        return ret;
}

# tcmu_notify_conn_lost()
{
...
        if (!tcmu_add_dev_to_recovery_list(dev))
                rdev->flags |= TCMUR_DEV_FLAG_IN_RECOVERY;
...
}

2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_notify_conn_lost:166 block0: Handler connection lost (lock state 1)
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_add_dev_to_recovery_list:327 block0: Could not find enabled port.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.334 5466 [ERROR] tcmu_rbd_handle_timedout_cmd:655 block0: Timing out cmd.
2017-09-08 10:01:06.335 5466 [DEBUG] handle_netlink:210 : cmd 2. Got header version 2. Supported 2.
2017-09-08 10:01:06.335 5466 [DEBUG] tcmu_cancel_recovery:135 block0: Waiting on recovery thread

Reply:
STOPPING checking at the beginning of tcmu_notify_conn_lost is needed, because code path
tcmu_rbd_handle_timedout_cmd -> tcmu_notify_conn_lost -> tcmu_add_dev_to_recovery_list -> tcmu_get_alua_grps will access the configfs dir resource which is holded by kernel in configfs_remove.
If tcmu_rbd_handle_timedout_cmd is blocked, in tcmulib_cmd->done() will never get a chance to clear tracked_aio_ops, which will cause deadlock in the following code path:

        cleanup_io_work_queue_threads(dev);

        if (aio_wait_for_empty_queue(rdev))
                tcmu_dev_err(dev, "could not flush queue.\n");
        tcmulib_cleanup_cmdproc_thread(dev);

        cleanup_io_work_queue(dev, false);
        cleanup_aio_tracking(rdev);

Reply:
@mikechristie @zhuozh 

With this new fix I couldn't reproduce the stuck any more on my environment.

Reply:
@zhuozh Nice debugging!

Reply:
For the final patch could you break up the check so it is in the 80 col limit we have been using?
if ((rdev->flags & TCMUR_DEV_FLAG_STOPPING) ||
    (rdev->flags & TCMUR_DEV_FLAG_IN_RECOVERY))

and add a comment about the issue, so new devs that do not do kernel work will know that we cannot change the behavior due to the kernel and not anything runner is doing.

Reply:
@mikechristie Thanks mike, PR updated by following your comments.